c/cert/test/rules/MEM30-C/test.c

```diff
@@ -0,0 +1,26 @@
+#include <stdlib.h>
```

Comment:
more of a question - but right now we definitely cover "dereferences after free" but what about the part of the rule that says "using it as an operand of an arithmetic operation, type casting it, and using it as the right-hand side of an assignment" 

do we want to also cover some of these cases? (I dont quite understand why stuff like pointer arithmetic (without dereference) would be bad per say - if it doesnt access the freed memory - but maybe a case of confusing intention?) 
```
  void test_freed(char *input) {
  char *buf = (char *)malloc(strlen(input) + 1);
  free(buf);

  char c0 = *buf; // NON_COMPLIANT

  char *newbuf;
  newbuf = buf; // NON_COMPLIANT? - RHS of an assignment
  char c1 = *newbuf;

  *buf = 'a'; // this is reported but then should the deref 2 lines down also be?
  // or is that getting into more of access unintialized memory...
  buf++;
  char c2 = *buf; // NON_COMPLIANT?

  double *d;
  *d = 0.9;
  free(d);
  char *p = (char *)&d; // NON_COMPLIANT? - typecast

  char *buf2 = (char *)malloc(strlen(input) + 1);
  free(buf2);
  free(buf2);       // NON_COMPLIANT
  char c02 = *buf2; // NON_COMPLIANT
}
```

Reply:
also, I wrote this then saw that its pretty much an import - maybe we can just add the extra testcases/or an implementation note, if you think the rule is making sense with the list of things to not do

Reply:
Thanks for noticing this! I did not extend the implementation to cover those additional cases as I interpreted "dereference" as "access". Looking back, you're right, and my interpretation is wrong. Here is an excerpt from a C99 Rationale document (https://www.open-std.org/jtc1/sc22/wg14/www/C99RationaleV5.10.pdf):

> An invalid pointer might be created in several ways. An arbitrary value can be assigned (via a cast) to a pointer variable. (This could even create a valid pointer, depending on the value.) A pointer to an object becomes invalid if the memory containing the object is deallocated or moved by realloc. Pointer arithmetic can produce pointers outside the range of an array.

> Regardless how an invalid pointer is created, any use of it yields undefined behavior. Even
assignment, comparison with a null pointer constant, or comparison with itself, might on some
systems result in an exception.

> Consider a hypothetical segmented architecture on which pointers comprise a segment descriptor
and an offset. Suppose that segments are relatively small so that large arrays are allocated in
multiple segments. While the segments are valid (allocated, mapped to real memory), the
hardware, operating system, or C implementation can make these multiple segments behave like
a single object: pointer arithmetic and relational operators use the defined mapping to impose
30 the proper order on the elements of the array. Once the memory is deallocated, the mapping is
no longer guaranteed to exist. Use of the segment descriptor might now cause an exception, or
the hardware addressing logic might return meaningless data. 

I will modify the query and test to model those additional test-cases.

Reply:
Changes made.

Reply:
lgtm!